Lombiq.HelpfulLibraries/Libraries/Mvc/ControllerExtensions.cs

```diff
@@ -25,5 +30,32 @@ public static class ControllerExtensions
         /// </summary>
         public static RedirectResult RedirectToContentDisplay(this Controller controller, IContent content) =>
             controller.Redirect(controller.Url.DisplayContentItem(content));
+
+        /// <summary>
+        /// Similar to <c>controller.Json(data)</c>, but catches any exception in the <paramref name="dataFactory"/> and
+        /// if one happens returns a JSON with the <c>error</c> property. If run from a local dev machine the
+        /// <c>data</c> property is also filled with the exception string.
+        /// </summary>
+        public static async Task<JsonResult> SafeJsonAsync<T>(this Controller controller, Func<Task<T>> dataFactory)
+        {
+            try
+            {
+                return controller.Json(await dataFactory());
+            }
+            catch (Exception exception)

```

Comment:
It should be more specific. Catching an exception that's thrown for an invalid argument is fine, but there could be other issues we want to actually see in the logs. Not to mention fatal exceptions like `OutOfMemoryException`. I'm more concerned about the latter, so if you don't have a better idea, let's just check if the exception `.IsFatal()` or not.

Reply:
This is necessary and I have been meaning to introduce something like it for a while anyway. If the client side expects a JSON response, then letting the exception tumble through and returning an HTML result (such as the Not Found error page) will only break the JS on the client page. So it should be caught for sure, but we can check if the exception is fatal and if so, do a `_logger.LogError`.

Reply:
I agree with you, but if there is an exception then it's a serious issue on its own, there shouldn't be any. If it crashes the front-end app, then so be it. If the exception represents a validation error, then it's a different story. I'll ping you @Piedone because we had a similar discussion about it a few years ago and I'm interested in your thoughts.

Reply:
> I agree with you, but if there is an exception then it's a serious issue on its own, there shouldn't be any.

Not all exceptions are serious, If you prefer, I can check `exception is not InvalidOperationException` (or `ValidationException`?) instead of `exception.IsFatal()`.

> If it crashes the front-end app, then so be it.

I strongly disagree. This attitude will lead to a bad UX and client friction.

Reply:
`IsFatal()` is needed for all base `Exception` catches.

Reply:
OK, let's check `IsFatal()` only.